src/ejabberd_c2s.erl

```diff
@@ -1133,29 +1135,21 @@ handle_incoming_message({send_filtered, Feature, From, To, Packet}, StateName, S
             ejabberd_router:route(From, To, FinalPacket),
             fsm_next_state(StateName, StateData)
     end;
-handle_incoming_message({broadcast, Type, From, Packet}, StateName, StateData) ->
-    %% this version is used only by mod_pubsub, which does things the old way
-    Recipients = mongoose_hooks:c2s_broadcast_recipients(StateData#state.server,
-                                                        [],
-                                                        StateData, Type, From, Packet),
-    lists:foreach(fun(USR) -> ejabberd_router:route(From, jid:make(USR), Packet) end,
-        lists:usort(Recipients)),
-    fsm_next_state(StateName, StateData);
-handle_incoming_message({call_info_handler, Tag, Data}, StateName, StateData) ->
-    NStateData =
-    case ejabberd_c2s_info_handler:get_for(Tag, StateData) of
-        {Tag, HandlerState} ->
-            case ejabberd_c2s_info_handler:safe_call(Tag, Data, HandlerState, StateData) of
-                {error, _} ->
-                    StateData;
-                NStateData0 ->
-                    NStateData0
-            end;
-        no_handler ->
-            ?INFO_MSG("event=no_info_handler tag=~p host=~p", [Tag, ejabberd_c2s_state:server(StateData)]),
-            StateData
-    end,
-    fsm_next_state(StateName, NStateData);
+handle_incoming_message({call_remote_hook, Tag, Args}, StateName, StateData) ->
+    Host = ejabberd_c2s_state:server(StateData),
+    HandlerState = maps:get(Tag, StateData#state.handlers, empty_state),
+    case mongoose_hooks:c2s_remote_hook_call(Host, Tag, Args, HandlerState, StateData) of
+        {error, E} ->
+            ?ERROR_MSG("event=custom_c2s_hook_handler_error tag=~p,data=~p "
+                       "extra=~p reason=~p",
+                       [Tag, Args, HandlerState, E]),
+            fsm_next_state(StateName, StateData);
+        HandlerState ->
+            fsm_next_state(StateName, StateData);
+        NewHandlerState ->
+            NewStates = maps:put(Tag, NewHandlerState, StateData#state.handlers),
+            fsm_next_state(StateName, StateData#state{handlers = NewStates})
```

Comment:
It should be fast enough for very small handler states. I also think it's optimized if NewHandlerState and HandlerState are on the same address.

I was thinking about allowing `same` or `skip` as a result of a hook. But it would break RARE cases when we need two hook handlers to work with the same handler state (for example, one is for just observing changes in hooks).

Alternatively, we can pass something like `{Updated :: true, HandlerState}`  as an acc in this hook. Or `{updated, HandlerState}` | `{same, HandlerState}` (we don't need to use boolean, where we can use domain specific atoms).

Reply:
Also, if we want to optimize it really hard (TM), we can use separate hooks for each hook handler.

We can also introduce concept of c2s events. Where each event would be a separate hook.
And to call something we just ask c2s process to emit specific event (i.e. to call a hook).

And hooks would be:

* c2s_ping_timeout
* c2s_ping_init
* c2s_ping_send


It would remove function clauses logic from each handler.

Reply:
> Alternatively, we can pass something like `{Updated :: true, HandlerState}` as an acc in this hook. Or `{updated, HandlerState}` | `{same, HandlerState}` (we don't need to use boolean, where we can use domain specific atoms).

I thought about something like this, though I don't know enough about Erlang internals to be able to judge if would be a real improvement or not - while most certainly some people would complain about adding complexity etc.

Reply:
> if we want to optimize it really hard (TM), we can use separate hooks for each hook handler.
> We can also introduce concept of c2s events. Where each event would be a separate hook.

In other words, we can stay with the current approach of c2s just calling specific hooks and modules registering specific handlers. That's what's already happening, although there are few of these events called via hooks (for example `privacy_check_packet`).

Reply:
> And hooks would be:
> 
>     * c2s_ping_timeout
> 
>     * c2s_ping_init
> 
>     * c2s_ping_send

i would rather want to have c2s_ping hook with event parameter (init/send/timeout)